File: lambda/save-draft/lambda_function.js
const AWS = require('aws-sdk');

const dynamodb = new AWS.DynamoDB.DocumentClient({ region: 'us-east-1' });
const s3 = new AWS.S3({ region: 'ap-south-2' });

const DYNAMODB_TABLE = process.env.DYNAMODB_TABLE || 'vmail-emails';
const S3_BUCKET = process.env.S3_BUCKET || 'vmail-emails-059409992687';

exports.handler = async (event, context) => {
  console.log('Save draft handler started');
  try {
    // Extract user info from Cognito authorizer
    const userId = event.requestContext.authorizer.claims.sub;
    const userEmail = event.requestContext.authorizer.claims.email;
    console.log('User:', userEmail, 'UserId:', userId);

    // Parse request body
    const body = JSON.parse(event.body);
    const { to, subject, body: emailBody, cc = [], bcc = [], attachments = [], draftId } = body;
    console.log('Saving draft for:', to, 'Subject:', subject);

    // Use provided draftId or generate new one
    const emailId = draftId || `${userId}-draft-${Date.now()}`;
    const timestamp = new Date().toISOString();

    // Store full email in S3
    const s3Key = `drafts/${userId}/${emailId}.json`;
    await s3.putObject({
      Bucket: S3_BUCKET,
      Key: s3Key,
      Body: JSON.stringify({
        subject,
        body: emailBody,
        from: userEmail,
        to: Array.isArray(to) ? to : [to],
        cc,
        bcc,
        attachments,
        timestamp
      }),
      ContentType: 'application/json'
    }).promise();

    // Store metadata in DynamoDB
    console.log('Storing draft in DynamoDB');
    await dynamodb.put({
      TableName: DYNAMODB_TABLE,
      Item: {
        emailId,
        userId,
        from: userEmail,
        to: Array.isArray(to) ? to : [to],
        cc,
        bcc,
        subject,
        preview: emailBody.replace(/<[^>]*>/g, '').substring(0, 100),
        timestamp,
        folder: 'drafts',
        read: true,
        starred: false,
        hasAttachments: attachments.length > 0,
        s3Key,
        isDraft: true
      }
    }).promise();
    console.log('Draft saved successfully');

    return {
      statusCode: 200,
      headers: getCorsHeaders(),
      body: JSON.stringify({
        message: 'Draft saved successfully',
        emailId,
        draftId: emailId
      })
    };
  } catch (error) {
    console.error('Error saving draft:', error);
    return {
      statusCode: 500,
      headers: getCorsHeaders(),
      body: JSON.stringify({
        message: `Error saving draft: ${error.message}`
      })
    };
  }
};

function getCorsHeaders() {
  return {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type,Authorization',
    'Access-Control-Allow-Methods': 'OPTIONS,POST,GET,PUT,DELETE'
  };
}

File: lambda/mark-starred/lambda_function.py
import json
import boto3
import os

# Initialize AWS clients
dynamodb = boto3.resource('dynamodb', region_name=os.environ.get('AWS_REGION', 'us-east-1'))

# Environment variables
DYNAMODB_TABLE = os.environ.get('DYNAMODB_TABLE', 'vmail-emails')

def lambda_handler(event, context):
    """
    Lambda function to toggle star on an email
    """
    try:
        # Extract user info from Cognito authorizer
        user_id = event['requestContext']['authorizer']['claims']['sub']

        # Get email ID from path parameters
        email_id = event['pathParameters']['emailId']

        # Parse request body to get starred status
        body = json.loads(event.get('body', '{}'))
        starred = body.get('starred', True)

        # Get email metadata from DynamoDB
        table = dynamodb.Table(DYNAMODB_TABLE)
        response = table.get_item(Key={'emailId': email_id})

        if 'Item' not in response:
            return {
                'statusCode': 404,
                'headers': get_cors_headers(),
                'body': json.dumps({'message': 'Email not found'})
            }

        email_metadata = response['Item']

        # Verify user has access to this email
        if email_metadata.get('userId') != user_id:
            return {
                'statusCode': 403,
                'headers': get_cors_headers(),
                'body': json.dumps({'message': 'Access denied'})
            }

        # Toggle starred status
        table.update_item(
            Key={'emailId': email_id},
            UpdateExpression='SET starred = :starred',
            ExpressionAttributeValues={':starred': starred}
        )

        return {
            'statusCode': 200,
            'headers': get_cors_headers(),
            'body': json.dumps({
                'message': 'Email starred status updated',
                'emailId': email_id,
                'starred': starred
            })
        }

    except Exception as e:
        print(f"Error marking email as starred: {str(e)}")
        return {
            'statusCode': 500,
            'headers': get_cors_headers(),
            'body': json.dumps({'message': f'Error: {str(e)}'})
        }

def get_cors_headers():
    return {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Content-Type,Authorization',
        'Access-Control-Allow-Methods': 'OPTIONS,POST,GET,PUT,DELETE'
    }
